### PR TITLE
Bug Fix #2585

### LINKED ISSUE #2585: Crash in SearchImageFragment
**Summary:** 
The Observable in SearchImageFragment has not been disposed as per the fragment lifecycle and because of which the app crashes (along with memory leaks), when RxCallbacks are received after the fragment is stopped/detached

**System logs:**

```
USER_COMMENT=French locale

ANDROID_VERSION=7.1.2
APP_VERSION_NAME=2.10.1.260
BRAND=google
PHONE_MODEL=Nexus 5
CUSTOM_DATA=
STACK_TRACE=io.reactivex.exceptions.OnErrorNotImplementedException: Attempt to invoke virtual method 'void android.view.View.setVisibility(int)' on a null object reference
at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:67)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:200)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:119)

at android.os.Handler.handleCallback(Handler.java:751)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:154)
at android.app.ActivityThread.main(ActivityThread.java:6186)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.View.setVisibility(int)' on a null object reference
at fr.free.nrw.commons.explore.images.SearchImageFragment.handleNoInternet(SearchImageFragment.java:233)
at fr.free.nrw.commons.explore.images.SearchImageFragment.updateImageList(SearchImageFragment.java:136)
at fr.free.nrw.commons.explore.SearchActivity.lambda$setTabs$1$SearchActivity(SearchActivity.java:112)
at fr.free.nrw.commons.explore.SearchActivity$$Lambda$1.accept(Unknown Source)
at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
... 10 more
java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.View.setVisibility(int)' on a null object reference
at fr.free.nrw.commons.explore.images.SearchImageFragment.handleNoInternet(SearchImageFragment.java:233)
at fr.free.nrw.commons.explore.images.SearchImageFragment.updateImageList(SearchImageFragment.java:136)
at fr.free.nrw.commons.explore.SearchActivity.lambda$setTabs$1$SearchActivity(SearchActivity.java:112)
at fr.free.nrw.commons.explore.SearchActivity$$Lambda$1.accept(Unknown Source)
at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:200)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:119)

at android.os.Handler.handleCallback(Handler.java:751)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:154)
at android.app.ActivityThread.main(ActivityThread.java:6186)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)  ```

**Device and Android version:** 
APP_VERSION_NAME=2.10.1.260
BRAND=google
PHONE_MODEL=Nexus 5

 
**Commons app version:** 
2.10.1.260

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by ashishkumar468**
**Description (required)**
Public methods are exposed to update ui of fragment from activity, and the methods are not lifecycle aware, added associated checks for ui in active state and null checks on views

Fixes  #2585 Crash in SearchImageFragment 

What changes did you make and why?
* Added null checks on view in SearchImageFragment when updating views from external sources
* Disposed the disposables in SearchActivity and SearchImageFragment when no longer in active lifecycle

**Tests performed (required)**

Tested { ProdDebug} on {Pixel 2} with API level {28}.

### CODE REVIEW

File: app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java [hunk 1]
```diff
@@ -98,28 +97,36 @@ public void setTabs() {
...
+            .debounce(500, TimeUnit.MILLISECONDS)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(query -> {
+                this.query = query.toString();
+                //update image list
+                if (!TextUtils.isEmpty(query)) {
+                    viewPager.setVisibility(View.VISIBLE);
+                    tabLayout.setVisibility(View.VISIBLE);
+                    searchHistoryContainer.setVisibility(View.GONE);
+                    if (searchImageFragment != null
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java]: Use `FragmentUtils:isFragmentUIActive`

File: app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java [hunk 2]
```diff
@@ -98,28 +97,36 @@ public void setTabs() {
...
+                        && searchImageFragment.isAdded()
+                        && !searchImageFragment.isDetached()) {
+                        searchImageFragment.updateImageList(query.toString());
+                    }
+
+                    if (searchCategoryFragment != null
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java]: Use `FragmentUtils:isFragmentUIActive`

File: app/src/main/java/fr/free/nrw/commons/explore/images/SearchImageFragment.java [hunk 3]
```diff
@@ -230,8 +229,15 @@ private void initErrorView() {
...
-        progressBar.setVisibility(GONE);
-        ViewUtil.showShortSnackbar(imagesRecyclerView, R.string.no_internet);
+        if (null
+            != getView()) {//We have exposed public methods to update our ui, we will have to add null checks until we make this lifecycle aware
+            if (null != progressBar) {
+                progressBar.setVisibility(GONE);
+            }
+            ViewUtil.showShortSnackbar(imagesRecyclerView, R.string.no_internet);
+        } else {
+            Timber.d(TAG + ": attempt to update fragment ui after its view was destroyed");
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/explore/images/SearchImageFragment.java]: `Timber` automatically adds the `TAG`. No need to append it explicitly.

**Comment by neslihanturan** [comment]
Thanks @ashishkumar468 , I couldn't reproduce any mem leaks on search fragment after your fixes. Seems like it works:)

**Pull request merged**